### PR TITLE
fix(build): restore the exec bit on the single-test path too (#4205, #4168)

### LINKED ISSUE #4168: bash test --cpp --clean leaves runner executables non-executable, faking a 372-test red
## Summary

`bash test --cpp --clean` leaves the quick build tree unusable: both runner executables are produced **without their executable bit**, so every subsequent test fails to spawn. The harness then reports a full-suite failure that is indistinguishable from a catastrophic regression.

Hit while validating an unrelated change; I initially misread it as my own breakage.

## Evidence

After `bash test --cpp --clean`:

| File | Mode |
|---|---|
| `.build/meson-quick/tests/runner` | `-rw-r--r--` :x: |
| `.build/meson-quick/examples/example_runner` | `-rw-r--r--` :x: |
| `.build/meson-debug/tests/runner` | `-rwxr-xr-x` :white_check_mark: |

Tests are built as `.so` DLLs loaded by `runner` (`tests/meson.build:85`). With `runner` non-executable, every one of the 372 tests fails:

```
[1/372] x example-AdafruitBridge FAILED
[MESON] Test execution error: failed to spawn process: Permission denied
(os error 13)
```

## Why this is worse than a one-off

1. **The bad state persists.** Every later `bash test --cpp` inherits it and dies at ~4s with `[MESON] Tests failed with return code -9`, plus `Could not find detailed error information in testlog.txt`. Nothing points at file permissions.
2. **The harness advice hides it.** On failure it prints `AGENT ACTION REQUIRED: RE-RUN FAILED TESTS IN DEBUG`. Debug mode uses `.build/meson-debug`, whose runner kept its bit, so the debug re-run *passes* — which reads as flakiness rather than a broken tree.
3. `-9` looks like an OOM kill. It is not: 97 GiB free and no kernel OOM record.

## Repro

```bash
bash test --cpp --clean     # 372 failures, "Permission denied (os error 13)"
bash test --cpp             # now fails at ~4s with return code -9
```

## Workaround

```bash
chmod +x .build/meson-quick/tests/runner .build/meson-quick/examples/example_runner
bash test --cpp             # 372/372 pass
```

Confirmed: immediately after the chmod the suite reported `All tests passed (372/372 unit + examples in 18.39s)` with no other change.

## Suspected cause

Likely the link/cache path rather than the build generator itself, which normally sets the mode on link. zccache warned throughout these runs:

```
WARNING: 100.0% of lookups returned cold_skip despite a populated depgraph on disk;
the warm-cache classifier may be regressed (see #320 / #796 SI-2)
```

A cache restore that writes the artifact without preserving file mode would produce exactly this. Worth checking whether zccache serves the runner link output and, if so, whether it round-trips mode bits.

## Suggested fixes

- Preserve/restore the executable bit wherever the runner link output is cached or copied.
- Assert the runner mode (or apply `chmod +x`) after build, so a bad cache entry cannot poison the tree.
- Have the test harness detect a non-executable runner and say so, instead of surfacing 372 spawn failures and pointing at debug mode.

## Environment

NixOS, Linux 6.18.48, x86_64. FastLED master @ `9de01d3547`.

Found during work on #3713.

Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01KkufoNxfnNRU9psT3R9F51

### LINKED ISSUE #4205: build: link wrapper strips the exec bit from test executables on every relink, causing os error 13 with no diagnostic
## Symptom

`bash test --cpp` fails with all tests unable to start:

```
[1/383] ✗ cled_controller FAILED
  [MESON] Test execution error: failed to spawn process: Permission denied (os error 13)
...
[MESON] ❌ Some tests failed (0/383 unit)
```

Sometimes it presents instead as a bare, reasonless failure:

```
ERROR: Unhandled python OSError. This is probably not a Meson bug, but an issue with your build environment.
[MESON] ⚠️  Test failed but could not extract detailed error information
[MESON] ❌ Tests failed with return code 13
Ok:  0    Fail:  0
```

This is the recurring symptom behind #4168. It is not stale state left by any particular merge — it reproduces on demand.

## Cause

Freshly linked **plain executables** come out without the exec bit. Shared objects are unaffected. Local umask is a normal `0022`.

After one build:
```
-rw-r--r--  1 niteris users    53920 tests/runner                      <- no +x
-rwxr-xr-x  1 niteris users    95320 tests/cycle_type.so               <- fine
-rw-r--r--  1 niteris users     6384 tests/color_profile_tiny_layout   <- no +x
```

`tests/runner` is the process the driver spawns, so when it loses the bit every single test fails to start. `color_profile_tiny_layout` is regenerated on **every** build, so it re-breaks continuously.

## Why it is hard to diagnose

Two things hide it:

1. `ci/meson/test_execution.py` streams the child's output but only echoes lines containing `FAILED` or `ERROR`, so the underlying traceback never reaches the user. You get `Ok: 0 Fail: 0` and "could not extract detailed error information". I had to bypass the wrapper to see the real error:
   ```
   PermissionError: [Errno 13] Permission denied:
     '.build/meson-quick/tests/color_profile_tiny_runtime'
   ```
2. `chmod +x` then re-running does not reliably help, because the build step inside `bash test` regenerates the affected file and strips the bit again before tests run. It only sticks if the build happens to be fully up to date.

The failure also triggers the "RE-RUN FAILED TESTS IN DEBUG MODE" banner, which is misleading here — sanitizers cannot instrument a process that will not spawn.

## Suggested fixes

- Have the link/deploy step (`ZCCACHE_LINK_DEPLOY_CMD` / `clang-tool-chain-libdeploy`) preserve the mode of what it emits, or `chmod +x` its output. That is the actual bug.
- Independently, make the test wrapper surface the child's stderr on a non-zero exit rather than filtering to FAILED/ERROR. The filtering is what turned a one-line `PermissionError` into an unexplained `Ok: 0 Fail: 0`.
- Consider suppressing the debug-mode banner when failures are spawn errors rather than test failures.

## Environment

NixOS, umask 0022, `nofile` 524288. Reproduced repeatedly across sessions today; each full relink reintroduces it.

### PR DESCRIPTION
**Pull request by zackees**
Closes #4205. Closes #4168.

## The half that was missed

#4213 added `restore_executable_bits` and wired it into two places: the
streaming compile path and `test_execution.py`, immediately before `meson test`
spawns the runner. That covers `bash test --cpp`.

It does not cover `bash test <name>`. That takes a third route —
`run_direct_test` in `ci/meson/sequential_runner.py` locates the runner and
spawns it directly — and that module never imported the function. A runner
zccache had served from its link cache stayed at `0644`, and the run died with

```
[MESON] Test execution failed with exception: failed to spawn process: Permission denied (os error 13)
```

which names neither the file nor the mode. That is #4168's complaint verbatim,
and it is still reproducible on `master` with #4213 merged:

```console
$ chmod -x .build/meson-quick/tests/runner
$ bash test cled_controller --cpp
3.88 [MESON] Test execution failed with exception: failed to spawn process: Permission denied (os error 13)
$ ls -l .build/meson-quick/tests/runner
-rw-r--r--  ...   # untouched
```

With this change, the same two commands:

```console
8.64 [BUILD] ⚡ Restored 1 executable bit(s) (zccache link-cache fix)
8.66 ✅ All tests passed (1/1 in 0.01s, build: 7.9s)
-rwxr-xr-x  ...
```

## Placement

The call goes *after* `_resolve_test_command`, not before it. That function
relinks the runner when it is missing, and a relink is exactly the moment the
bit goes away — restoring first would race the thing it is repairing.

## The test

`ci/tests/test_direct_test_exec_bit.py` drives the real `run_direct_test`
against a real executable, rather than asserting that a particular function
gets called. Two mutations were run against it:

| mutation | result |
|---|---|
| drop the `restore_executable_bits` call | fails with the original `os error 13` |
| move it after the spawn | fails with the original `os error 13` |

A second case pins that an already-executable runner keeps its exact mode, so
the restore cannot quietly widen permissions.

The stand-in runner is a copy of `sh`, not `true`: `restore_executable_bits`
only touches files with a native-executable magic number (so a text stub would
be skipped for the right reason and prove nothing), and a multi-call coreutils
binary dispatches on `argv[0]` and refuses to run under the name `runner`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MLhWkMfzLjrnLTDMBE6Fj9

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed an issue where running a single test could fail with a “Permission denied” error when the test executable lacked executable permissions.
  - Test runners now launch successfully after permissions are restored, while existing executable permissions remain unchanged.

### CODE REVIEW

File: ci/meson/sequential_runner.py [hunk 1]
```diff
@@ -465,6 +466,14 @@ def run_direct_test(ctx: DirectTestContext) -> MesonTestResult:
             num_tests_failed=0,
         )
 
+    # After resolution, not before: `_resolve_test_command` may relink the
+    # runner, and a link zccache serves from its cache comes back without the
+    # executable bit (FastLED#4205). `test_execution.py` restores it on the
+    # `meson test` path; this is the other way a test binary gets spawned, and
+    # it had no such call -- so a single named test died on
+    # `Permission denied (os error 13)` with nothing pointing at the mode.
+    restore_executable_bits(ctx.build_dir)
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/meson/sequential_runner.py:475]: Confirmed and fixed. `restore_executable_bits` scanned `tests` and `examples` without recursing, so `tests/profile` -- which `_resolve_test_command` prefers, spawning the binary directly with no runner -- was walked straight past.

The directory is named rather than recursed into: the scan deliberately looks only where linked executables land, and walking the whole build tree would touch object directories for no reason.

`ci/tests/test_direct_test_exec_bit.py` gains a case that lays down a profile executable at mode 0644 and drives the real `run_direct_test` against it. Verified by mutation: dropping `tests/profile` from the scan fails it with the original `failed to spawn process: Permission denied (os error 13)`.

**Comment by zackees** [comment]
Fixed. The finding is correct: `_resolve_test_command()` can select `build_dir/tests/profile/<name>`, the restore scan covered only `("tests", "examples")`, and `iterdir()` is not recursive — so `tests/profile` was skipped entirely and a cache-restored profile binary stayed at `0644`, failing to spawn with `PermissionError`. That is the same defect this function exists to prevent, one directory over.

`tests/profile` is now in the scan list, with a comment naming why it is separate rather than implied by `tests`.

Added a test that creates a `0644` ELF under `tests/profile` and asserts it is repaired, verified by reverting the directory list — the test fails, and passes again when restored. 7 collected (was 6), 7 passed.

Worth noting the general shape, since this is the third time it has come up in this area: the failure mode is always a *silent skip* — a directory not scanned, a row not checked, a `Path.is_dir()` returning `False` for something it cannot stat. Each one turns "we repaired everything" into "we repaired what we happened to look at", and the symptom surfaces much later as an opaque `PermissionError` with no connection to the cause.

**Comment by zackees** [comment]
Correction to my previous comment: when I posted it the fix was committed locally but **not pushed** — the push was rejected because the same fix had already landed on this branch as `cc176dfea3`. So the code change was already there and my claim that I had fixed it was premature.

Resolved now. The code change is identical on both sides (adding `tests/profile` to the scan list), so I kept the existing comment and dropped my duplicate. What remains from my side is additive: a regression test, which this branch did not have for this path.

```python
def test_restore_executable_bits_covers_profile_tests(tmp_path):
    (tmp_path / "tests" / "profile").mkdir(parents=True)
    profile_binary = _elf(tmp_path / "tests" / "profile" / "profile_hot", mode=0o644)
    assert restore_executable_bits(tmp_path) == 1
    assert profile_binary.stat().st_mode & 0o111
```

Verified by reverting the directory list: the test fails, and passes again when restored. 7 collected (was 6), 7 passed.

Without that test, the next reorganisation of the scan list silently reopens the hole — which matters here because the symptom is an opaque `PermissionError` at spawn time with no connection to the cause.
